simulate/sam.go

```diff
@@ -13,21 +13,50 @@ import (
 	"github.com/vertgenlab/gonomics/sam"
 )
 
-// IlluminaPairedSam generates a pair of sam reads randomly distributed across the input ref.
+// IlluminaPairedSam generates pairs of sam reads randomly distributed across the input DNA sequence.
+// The inputs are the name of the input DNA sequence, the sequence itself, the number of read pairs to generate, the length of each read,
+// the average fragment size, the standard deviation of fragment sizes, the error rate where a base in
+// the read will not match the input DNA, a numbers.binomialAlias that is used to speed up calculations, and
+// output file handles for sam, bam, and a bool denoting if bam (or sam) should be the output.
+// Whichever handle (sam or bam) is not being used can be nil.
 func IlluminaPairedSam(refName string, ref []dna.Base, numPairs, readLen, avgFragmentSize int, avgFragmentStdDev float64, flatErrorRate float64, ancientErrorRate float64, flatBinomialAlias numbers.BinomialAlias, ancientBinomialAlias numbers.BinomialAlias, geometricParam float64, out *fileio.EasyWriter, bw *sam.BamWriter, bamOutput bool, deaminationDistributionSlice []int) {
-	var fragmentSize, midpoint, startFor, startRev, endFor, endRev int
+	var fragmentSize, midpoint, startFor, endRev, newCapacity int
 	var currFor, currRev sam.Sam
+	var fragment []dna.Base = make([]dna.Base, 0, avgFragmentSize+int(5*avgFragmentStdDev)) //set initial fragment slice capacity to 5 deviations above average fragment size
+	var newFragment []dna.Base                                                              //this will be used if we run out of capacity in 'fragment'
+	if avgFragmentSize < readLen {
+		log.Fatalf("Error: average fragment size %v is less than read length %v.\n", avgFragmentSize, readLen)
```

Comment:
Can you help me better understand what "fragment" is? Is it a subset of the whole input DNA sequence (e.g. a region on Neanderthal chr1)? Is fragment used as a "unit" to generate matching ancient damage on both forward and reverse reads?

Why is it a fatal error if the average fragment size is less than read length? Is this something the user should take care to avoid before they feed in the inputs to this IlluminaPairedSam function?

Reply:
Thank you! I am approving and merging since I don't think the code needs to change, but I have 2 followup questions.

Now I understand that the capacity refers to the memory allocated to the fragment. Does fragmentSize represent the number of bases (int), and len(fragment) measures how much memory the fragment takes up (length of []dna.Base))?

My experience has been that when fragmentSize < readLength, full-length reads are still produced, but the bases after the fragment ends are all empty calls (e.g. G calls for certain chemistries). Do you experience truncated reads?